### ISSUE TITLE
[SR-15964] ArgumentParser LTO crashes linkers

### ISSUE THREAD

**Issue by keith** [apple/swift#58225]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-15964      |
   |Radar             | None         |
   |Original Reporter | @keith      |
   |Type              | Bug    |

<details>
  <summary>Environment</summary>

<https://github.com/apple/swift/commit/e3b271960dc491a1682b82bbe4c77360f790d177>  
This also repros with Xcode 13.2.1 and 13.3.0 RC 1

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       |     |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 30aec9c9489d7a67fe42931116a6a93e

  </details>

**Issue Description:**

When building swift-argument-parser with LTO enabled, ld64 (and lld on macOS) crash in the LTO logic. I believe this is because Swift emits invalid LLVM BC for this case. When stepping through LLVM in the debugger (with lld, ld64 might be different but is harder for me to debug) the issue is that a string gets loaded from the strtab with a size of 0, which hits an assertion. I traced the loading back to this function <https://github.com/llvm/llvm-project/blob/10e99eb7e4adbb8e2407e2f6ae3d2e6420d060c9/llvm/lib/Bitcode/Reader/BitcodeReader.cpp#L461-L469> and confirmed that in some cases \`Record\[1\]\` is 0 when seemingly it should always be non-0. I'm not sure what to look for in the bitcode or how to trace back the specific case, but maybe that would shed some more light on the issue.

The easiest way to reproduce this is to build with CMake (swiftpm has LTO support issues today):

``` none
cd /path/to/swift-argument-parser
mkdir build
cd build
cmake .. -G Ninja -DCMAKE_Swift_FLAGS='-lto=llvm-thin' -DCMAKE_Swift_COMPILER=/path/to/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/bin/swiftc
```

Here's a snippet of the ld64 crash report:

``` none
xception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000008
Exception Codes:       0x0000000000000001, 0x0000000000000008
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [78090]

VM Region Info: 0x8 is not in any region.  Bytes before following region: 4333355000
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                      10249c000-102644000    [ 1696K] r-x/r-x SM=COW  ...in/usr/bin/ld

Application Specific Information:
ld64-711
ld -dylib Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Completions/BashCompletionsGenerator.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Completions/CompletionsGenerator.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Completions/FishCompletionsGenerator.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Completions/ZshCompletionsGenerator.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Properties/Argument.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Properties/ArgumentHelp.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Properties/CompletionKind.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Properties/Errors.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Properties/Flag.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Properties/NameSpecification.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Properties/Option.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Properties/OptionGroup.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Types/CommandConfiguration.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Types/EnumerableFlag.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Types/ExpressibleByArgument.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Types/ParsableArguments.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Types/ParsableArgumentsValidation.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsable_Types/ParsableCommand.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsing/ArgumentDecoder.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.dir/Parsing/ArgumentDefinition.swift.o.bc Sources/ArgumentParser/CMakeFiles/ArgumentParser.d


Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libLTO.dylib                           0x107f6b2d4 llvm::DataLayout::getAlignment(llvm::Type*, bool) const + 24
1   libLTO.dylib                           0x107f69118 llvm::DataLayout::getABITypeAlign(llvm::Type*) const + 16
2   libLTO.dylib                           0x107f69118 llvm::DataLayout::getABITypeAlign(llvm::Type*) const + 16
3   libLTO.dylib                           0x1081c5c34 llvm::TargetLoweringBase::getABIAlignmentForCallingConv(llvm::Type*, llvm::DataLayout) const + 16
4   libLTO.dylib                           0x108430c84 llvm::TargetLowering::LowerCallTo(llvm::TargetLowering::CallLoweringInfo&) const + 2272
5   libLTO.dylib                           0x108445d64 llvm::SelectionDAGBuilder::lowerInvokable(llvm::TargetLowering::CallLoweringInfo&, llvm::BasicBlock const*) + 84
6   libLTO.dylib                           0x1084367c0 llvm::SelectionDAGBuilder::LowerCallTo(llvm::CallBase const&, llvm::SDValue, bool, llvm::BasicBlock const*, llvm::TargetLowering::PtrAuthInfo const*) + 1704
```

This repros with thin and full LTO, with optimizations and without. I singled out swift-argument-parser here because I have successfully produced LTO builds on tools without it as a dependency, and it itself has no dependencies so maybe that simplifies things.

**Comment by keith** [comment]
This _might_ be fixed on main, but I'm not 100% sure. It's quick finicky to test real world scenarios because of LTO mismatch issues with custom toolchains. For future readers LIBLTO_PATH likely needs to be set for libtool to be happy when using a toolchain, and Xcode's internal versions matter here. I think I got unlucky testing with a snapshot around a major Xcode release. I might try again in a few weeks

**Comment by keith** [comment]
Ok here's a more isolated repro case:

```bash
#!/bin/bash

rm -f ./*.swift.bc ./*.a ./output.json ./empty.swift ./something.swift

touch empty.swift
echo '
enum Foo {
    case bar
}
' > something.swift

echo '
{
 "something.swift": {
  "llvm-bc": "something.swift.bc"
  },
 "empty.swift": {
  "llvm-bc": "empty.swift.bc"
  }
}
' > output.json

swiftc \
  -emit-bc \
 something.swift \
 empty.swift \
 -target arm64-apple-macosx13.0 \
 -lto=llvm-thin \
 -module-name somemodule \
 -output-file-map output.json

ld \
 -arch arm64 \
 -bundle \
 -platform_version macos 13.0 13.0 \
 -o foo \
 something.swift.bc empty.swift.bc
```

This repros 100% for me on macOS with Swift 5.7*. It doesn't repro for me with the HEAD development snapshot from 12-10-22, so maybe that means it will be fixed in Swift 5.8.

**Comment by keith** [comment]
Seems like Xcode 14.3 beta 3 no longer has this crash
